File: api/models/note.py
"""
Pydantic models for Notes.
"""

from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from datetime import datetime


class AttachmentResponse(BaseModel):
    """Attachment data."""
    id: int
    file_path: str
    file_type: Optional[str] = None
    upload_date: Optional[datetime] = None


class NoteCreate(BaseModel):
    """Request model for creating a note."""
    title: str = Field(..., min_length=3, max_length=100)
    content: str = Field(default="")
    category: Optional[str] = None
    importance: int = Field(default=3, ge=1, le=5)
    color: Optional[str] = None
    tags: List[str] = Field(default_factory=list)


class NoteUpdate(BaseModel):
    """Request model for updating a note."""
    title: Optional[str] = Field(None, min_length=3, max_length=100)
    content: Optional[str] = None
    category: Optional[str] = None
    importance: Optional[int] = Field(None, ge=1, le=5)
    color: Optional[str] = None
    tags: Optional[List[str]] = None


class NoteResponse(BaseModel):
    """Response model for a single note."""
    id: int
    user_id: int
    title: str
    content: str
    category: Optional[str] = None
    importance: int = 3
    color: Optional[str] = None
    created_date: Optional[datetime] = None
    last_updated: Optional[datetime] = None
    archived: int = 0
    pinned: int = 0
    tags: List[str] = Field(default_factory=list)
    attachments: List[AttachmentResponse] = Field(default_factory=list)

    class Config:
        from_attributes = True


class NoteListResponse(BaseModel):
    """Response model for list of notes."""
    notes: List[NoteResponse]
    total: int


class NoteStatsResponse(BaseModel):
    """Response model for note statistics."""
    total: int
    pinned: int
    archived: int
    active: int
    by_category: Dict[str, int] = Field(default_factory=dict)


class TagListResponse(BaseModel):
    """Response model for list of tags."""
    tags: List[str]


class CategoryListResponse(BaseModel):
    """Response model for list of categories."""
    categories: List[str]

File: frontend/src/types/note.ts
export interface Attachment {
  id: number;
  file_path: string;
  file_type: string | null;
  original_filename?: string | null;
  file_size?: number | null;
  upload_date: string | null;
}

export interface Note {
  id: number;
  user_id: number;
  title: string;
  content: string;
  category: string | null;
  importance: number;
  color: string | null;
  created_date: string | null;
  last_updated: string | null;
  archived: number;
  pinned: number;
  tags: string[];
  attachments: Attachment[];
}

export interface NoteCreate {
  title: string;
  content: string;
  category?: string | null;
  importance?: number;
  color?: string | null;
  tags?: string[];
}

export interface NoteUpdate {
  title?: string;
  content?: string;
  category?: string | null;
  importance?: number;
  color?: string | null;
  tags?: string[];
}

export interface NoteListResponse {
  notes: Note[];
  total: number;
}

export interface NoteStats {
  total: number;
  pinned: number;
  archived: number;
  active: number;
  by_category: Record<string, number>;
}

export interface NoteFilter {
  query?: string;
  category?: string;
  tag?: string;
  importance?: number;
  archived?: boolean;
  pinned_only?: boolean;
}
